### ISSUE TITLE
M1 Mac support

### ISSUE THREAD

**Issue by simonholzapfel** [tensorflow/recommenders#463]
Tensorflow recommenders doesn't work on m1 mac

tensorflow: 2.8.0
tensorflow-recommenders: 0.6.0

I installed tensorflow using the following tutorial:
https://www.mrdbourke.com/setup-apple-m1-pro-and-m1-max-for-machine-learning-and-data-science/

After installing tensorflow recommenders i noticed that it doesn't work with metal gpu hardware acceleration.

<img width="1353" alt="image" src="https://user-images.githubusercontent.com/38250977/157823738-253cc88b-d9ec-4f49-82b8-af6edb728d38.png">

**Comment by maciejkula** [comment]
The error you show doesn't seem related to Applie silicon. It looks like you're passing Pandas dataframes in places where functions are expected.

**Comment by simonholzapfel** [comment]
I am sure it is, to test if something in my project is flawed, i checked [the movielens100k sample by tensorflow](https://www.tensorflow.org/recommenders/examples/quickstart) and it only works when i disable gpu hardware acceleration using `tf.config.set_visible_devices([], 'GPU')`.

**Comment by simonholzapfel** [comment]
Oh, I just noticed that I provided the wrong screenshot. You can now see the correct one.

**Comment by patrickorlando** [comment]
Hey @simonholzapfel,

Are you by any chance using Adagrad for your optimizer? Try changing it to Adam or SGD and see whether that resolves your issue.

**Comment by simonholzapfel** [comment]
Thank you, GPU hardware acceleration works using Adam and SGD, however it is slower than Adagrad using CPU only.
Is this because of Adam and SGD being computationally more complex or am I doing something wrong?

(Using StringLookups with huge amounts of values on M1 Pro with 8 CPU Cores and 14 GPU cores)

**Comment by patrickorlando** [comment]
SGD will have slower convergence, but definitely shouldn't have slower step speed. Adam does have some higher memory constraints but generally shouldn't slow down your step speed too much either.

If SGD takes longer per step on GPU then try the following.

String lookups have to happen on the CPU, and so if they are happening when you call the model, there is time where the GPU is idle whilst these are computed. Instead since they have no trainable weights you can run them in parallel as you prepare batches by placing these ops in your `tf.data` pipeline.

Basically you refactor your code to allow you to separate these steps during training, then package them together for inference.

This has provided higher GPU utilisation in my experience.

**Comment by maciejkula** [comment]
+1 for moving lookups to `tf.data` pipelines.
